libminifi/include/core/CachedValueValidator.h

```diff
@@ -0,0 +1,103 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NIFI_MINIFI_CPP_CACHEDVALUEVALIDATOR_H
+#define NIFI_MINIFI_CPP_CACHEDVALUEVALIDATOR_H
+
+#include "PropertyValidation.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace core {
+
+class CachedValueValidator{
+ public:
+  enum class Result {
+    FAILURE,
+    SUCCESS,
+    RECOMPUTE
+  };
+
+  CachedValueValidator() = default;
+  CachedValueValidator(const CachedValueValidator& other) : validator_(other.validator_) {}
+  CachedValueValidator(CachedValueValidator&& other) noexcept : validator_(std::move(other.validator_)) {}
```

Comment:
I generally like to take only what I use and making it as explicit as possible, instead of adjusting the effects after the fact, but if you think this would improve the readability I could make the change

Reply:
In my understanding, when I move construct something, I want the other object's content in the new object and preferably leave the other object in a somewhat valid state. (There is no clear consensus in the community about the last part as far as I'm aware.)
What happens here is we take the validator, but not the cached result. This way after a move, we are forced to redo the validation, and I assume this is somewhat expensive, otherwise we wouldn't need a cache. In the other object, we leave a moved-from validator with a validation result that no longer corresponds to the validator, and is therefore invalid.

Given the above, this seems less than ideal, but it's possible that I didn't fully understand your point and the code.

Reply:
the only owner of a `CachedValueValidator` instance should be a `PropertyValue`, ideally the move-constructor (and move-assignment) should be deleted, but since the `PropertyValue` has those defaulted that's not an option, when you move a `CachedValueValidator` into a `PropertyValue::validator_` you change the validator object in a value thus the cached validation result cannot be trusted anymore, the most important part is that such a `CachedValueValidator` should only be executed on its owner value